MdeModulePkg/Library/MemoryBucketLib/MemoryBucketLib.c

```diff
@@ -0,0 +1,447 @@
+/**@file
+Library for defining PEI memory buckets. This keeps track of the
+different buckets and the data for the associated HOB.
+For internal use only.
+
+Copyright (c) Microsoft Corporation.
+SPDX-License-Identifier: BSD-2-Clause-Patent
+**/
+
+#include <PiPei.h>
+
+#include <Library/DebugLib.h>
+#include <Library/MemoryBucketLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/HobLib.h>
+
+// Memory types being kept in buckets in PEI.  Currently only runtime types.
+EFI_MEMORY_TYPE  MemoryTypes[4] = {
+  EfiRuntimeServicesCode,
+  EfiRuntimeServicesData,
+  EfiACPIReclaimMemory,
+  EfiACPIMemoryNVS
+};
+
+// PEI memory bucket statistics.  Can be extended if necessary
+EFI_MEMORY_TYPE_STATISTICS  RuntimeMemoryStats[4] = {
+  { 0, MAX_ALLOC_ADDRESS, 0, 0, EfiRuntimeServicesCode, TRUE, TRUE  },   // EfiRuntimeServicesCode
+  { 0, MAX_ALLOC_ADDRESS, 0, 0, EfiRuntimeServicesData, TRUE, TRUE  },   // EfiRuntimeServicesData
+  { 0, MAX_ALLOC_ADDRESS, 0, 0, EfiACPIReclaimMemory,   TRUE, FALSE },   // EfiACPIReclaimMemory
+  { 0, MAX_ALLOC_ADDRESS, 0, 0, EfiACPIMemoryNVS,       TRUE, FALSE }    // EfiACPIMemoryNVS
+};
+
+EFI_PHYSICAL_ADDRESS  CurrentBucketTops[4] = {
+  MAX_ALLOC_ADDRESS, // EfiRuntimeServicesCode
+  MAX_ALLOC_ADDRESS, // EfiRuntimeServicesData
+  MAX_ALLOC_ADDRESS, // EfiACPIReclaimMemory
+  MAX_ALLOC_ADDRESS  // EfiACPIMemoryNVS
+};
+
+// Total number of memory buckets
+UINTN  NumberOfBuckets = 4;
+
+// Information storage for Hob
+PEI_MEMORY_BUCKET_INFORMATION  RuntimeBucketHob;
+
+// TRUE if PEI memory buckets are disabled
+BOOLEAN  MemoryBucketsDisabled = FALSE;
+
+// TRUE if we have initialized the runtime memory buckets.
+BOOLEAN  RuntimeMemInitialized = FALSE;
+
+/**
+  TODO
+**/
+VOID
+EFIAPI
+InitializeMemoryBucketSizes (
+  VOID
+  )
+{
+  RuntimeMemoryStats[0].NumberOfPages = FixedPcdGet8 (PcdPeiMemoryBucketRuntimeCode);

```

Comment:
An enum would be nice here. Haven't checked usage of index to bucket type elsewhere yet.

Reply:
It looks like you're using `MemoryTypeToIndex()` to do the conversion. But it's not being used in places like this leaving a gap between what it returns and what is being used here.

Reply:
You're suggesting to use another enum to represent the 0-3 values for easier readability?